Command argument splatting

```
let x = ["a", "b"] in
`echo $..x`

~> 

`echo a b`
```

```
(\x -> `echo $..x`) : Array String -> Cmd
```

* [x] Implementation
* [x] Documentation
  * Feature
  * Grammar

---

As part of #138, I changed command argument interpolation to use the same rules as string interpolation.

Variable interpolation in commands used to choose between expanding to a single argument or multiple arguments based on the variable's type:

```
(\x -> `echo $x`) : ToArgs a=> a -> Cmd
```

Interpolating an `Array String` would expand to multiple arguments:

```
let x = ["a", "b"] in
`echo $x`

~>

`echo a b`
```

Since #138, you can't interpolate variables of type `Array String` into commands. Only `String`s:

```
(\x -> `echo $x`) : String -> Cmd
```

I mostly did this by accident; my intent was to simplify the implementation of command syntax. I then realised that the explicitness of the approach could be helpful.

Should the result of

```
let x = ["a", "b"] in
`echo $x`
```

be 

```
`echo a b`
```

or

```
`echo "a b"`
```

or

```
`echo "[\"a\", \"b\"]"`
```

or something else?

Some of the options are more sensible than others, but the fact that I can *ask the question* also seems like a problem in itself.

When `Array String` can be substituted using `$`,

```
`echo $x`
```

and

```
`echo "$x"`
```

can mean different things. If `x : Array String`, then adding the double quotes changes the meaning of the command. I think it would be good if adding double quotes like this was *always* redundant, rather than "only being redundant when `x : String`". Substituting `String` and `Array String` with `$` reminds me a bit too much of current shells, in which [shellcheck recommends always quoting substitutions](https://www.shellcheck.net/wiki/SC2086) to avoid weirdness. `ipso` doesn't have that weirdness, but I think it'll be nice to move in the other direction.

--- a/examples/__skip.dhall
+++ b/examples/__skip.dhall
@@ -1,4 +1,3 @@
 -- { name = "test name", reason = "reason you're skipping it" }
-[ { name = "cmdInterpolation2", reason = "waiting for command argument splatting - https://github.com/LightAndLight/ipso/issues/346" }
-, { name = "cmdInterpolation3", reason = "waiting for command argument splatting - https://github.com/LightAndLight/ipso/issues/346" }
+[ 
 ] : List { name : Text, reason : Text }

--- a/examples/cmdInterpolation2.ipso
+++ b/examples/cmdInterpolation2.ipso
@@ -4,6 +4,6 @@ main =
     let a = "hello"
     println <| cmd.show `echo $a`
     let b = ["hello", "world"]
-    println <| cmd.show `echo $b`
+    println <| cmd.show `echo $..b`
     let c = ["hello", "beautiful world"]
-    println <| cmd.show `echo $c`
+    println <| cmd.show `echo $..c`

--- a/examples/cmdInterpolation3.ipso
+++ b/examples/cmdInterpolation3.ipso
@@ -1,5 +1,5 @@
 main : IO ()
 main =
   comp 
-    println <| cmd.show `echo ${ ["hello", "world"] }`
+    println <| cmd.show `echo $..{ ["hello", "world"] }`
     println <| cmd.show `echo ${ string.join " " ["hello", "world"] }`

--- a/ipso-core/src/lib.rs
+++ b/ipso-core/src/lib.rs
@@ -829,8 +829,9 @@ pub enum Binop {
 }
 
 #[derive(Debug, PartialEq, Eq, Clone)]
-pub struct CmdPart<E> {
-    pub value: Vec<StringPart<E>>,
+pub enum CmdPart<E> {
+    Arg(Vec<StringPart<E>>),
+    Args(E),
 }
 
 #[derive(Debug, PartialEq, Eq, Clone, Hash)]
@@ -1119,12 +1120,14 @@ impl Expr {
                 Expr::Cmd(parts) => Expr::Cmd(
                     parts
                         .iter()
-                        .map(|part| CmdPart {
-                            value: part
-                                .value
-                                .iter()
-                                .map(|string_part| string_part.map_expr(|expr| go(expr, f)))
-                                .collect(),
+                        .map(|cmd_part| match cmd_part {
+                            CmdPart::Arg(string_parts) => CmdPart::Arg(
+                                string_parts
+                                    .iter()
+                                    .map(|string_part| string_part.map_expr(|expr| go(expr, f)))
+                                    .collect(),
+                            ),
+                            CmdPart::Args(expr) => CmdPart::Args(go(expr, f)),
                         })
                         .collect(),
                 ),
@@ -1242,12 +1245,14 @@ impl Expr {
             Expr::Cmd(parts) => Expr::Cmd(
                 parts
                     .iter()
-                    .map(|part| CmdPart {
-                        value: part
-                            .value
-                            .iter()
-                            .map(|string_part| string_part.__instantiate(depth, val))
-                            .collect(),
+                    .map(|cmd_part| match cmd_part {
+                        CmdPart::Arg(string_parts) => CmdPart::Arg(
+                            string_parts
+                                .iter()
+                                .map(|string_part| string_part.__instantiate(depth, val))
+                                .collect(),
+                        ),
+                        CmdPart::Args(expr) => CmdPart::Args(expr.__instantiate(depth, val)),
                     })
                     .collect(),
             ),
@@ -1321,12 +1326,17 @@ impl Expr {
 
             Expr::Unit => Ok(()),
 
-            Expr::Cmd(parts) => parts.iter_mut().try_for_each(|part| {
-                part.value
-                    .iter_mut()
-                    .try_for_each(|string_part| string_part.subst_placeholder(f))?;
-                Ok(())
-            }),
+            Expr::Cmd(cmd_parts) => cmd_parts
+                .iter_mut()
+                .try_for_each(|cmd_part| match cmd_part {
+                    CmdPart::Arg(string_parts) => {
+                        string_parts
+                            .iter_mut()
+                            .try_for_each(|string_part| string_part.subst_placeholder(f))?;
+                        Ok(())
+                    }
+                    CmdPart::Args(expr) => expr.subst_placeholder(f),
+                }),
         }
     }
 
@@ -1408,15 +1418,17 @@ impl Expr {
                 bs.iter().map(|b| b.__abstract_evar(depth, ev)).collect(),
             ),
             Expr::Unit => Expr::Unit,
-            Expr::Cmd(parts) => Expr::Cmd(
-                parts
+            Expr::Cmd(cmd_parts) => Expr::Cmd(
+                cmd_parts
                     .iter()
-                    .map(|part| CmdPart {
-                        value: part
-                            .value
-                            .iter()
-                            .map(|string_part| string_part.__abstract_evar(depth, ev))
-                            .collect(),
+                    .map(|cmd_part| match cmd_part {
+                        CmdPart::Arg(string_parts) => CmdPart::Arg(
+                            string_parts
+                                .iter()
+                                .map(|string_part| string_part.__abstract_evar(depth, ev))
+                                .collect(),
+                        ),
+                        CmdPart::Args(expr) => CmdPart::Args(expr.__abstract_evar(depth, ev)),
                     })
                     .collect(),
             ),
@@ -1554,18 +1566,17 @@ impl Expr {
                         });
                         stack.push(a);
                     }
-                    Expr::Cmd(parts) => {
-                        parts.iter().rev().for_each(|cmd_part| {
-                            cmd_part
-                                .value
-                                .iter()
-                                .rev()
-                                .for_each(|string_part| match string_part {
+                    Expr::Cmd(cmd_parts) => {
+                        cmd_parts.iter().rev().for_each(|cmd_part| match cmd_part {
+                            CmdPart::Arg(string_parts) => string_parts.iter().rev().for_each(
+                                |string_part| match string_part {
                                     StringPart::String(_) => {}
                                     StringPart::Expr(expr) => {
                                         stack.push(expr);
                                     }
-                                })
+                                },
+                            ),
+                            CmdPart::Args(expr) => stack.push(expr),
                         });
                     }
                     Expr::EVar(a) => {

--- a/ipso-eval/src/closure_conversion/mod.rs
+++ b/ipso-eval/src/closure_conversion/mod.rs
@@ -359,8 +359,13 @@ pub fn convert_string_part(part: &StringPart<ipso_core::Expr>) -> ConvertResult<
     }
 }
 
-pub fn convert_cmd_part(part: &CmdPart<ipso_core::Expr>) -> ConvertResult<CmdPart<Expr>> {
-    convert_many(convert_string_part, &part.value).map(|value| CmdPart { value })
+pub fn convert_cmd_part(cmd_part: &CmdPart<ipso_core::Expr>) -> ConvertResult<CmdPart<Expr>> {
+    match cmd_part {
+        CmdPart::Arg(string_parts) => {
+            convert_many(convert_string_part, &string_parts).map(CmdPart::Arg)
+        }
+        CmdPart::Args(expr) => convert_expr(expr).map(CmdPart::Args),
+    }
 }
 
 pub fn convert_branch(branch: &Branch<ipso_core::Expr>) -> ConvertResult<Branch<Expr>> {

--- a/ipso-eval/src/lib.rs
+++ b/ipso-eval/src/lib.rs
@@ -2536,8 +2536,16 @@ where {
             Expr::Cmd(cmd_parts) => {
                 let mut new_cmd_parts: Vec<Rc<str>> = Vec::with_capacity(cmd_parts.len());
                 for cmd_part in cmd_parts {
-                    let new_part = self.eval_string_parts(env, &cmd_part.value);
-                    new_cmd_parts.push(Rc::from(new_part.as_str()));
+                    match cmd_part {
+                        core::CmdPart::Arg(string_parts) => {
+                            let new_part = self.eval_string_parts(env, string_parts);
+                            new_cmd_parts.push(Rc::from(new_part.as_str()));
+                        }
+                        core::CmdPart::Args(expr) => {
+                            let args = self.eval(env, expr).unpack_array();
+                            new_cmd_parts.extend(args.iter().map(|arg| arg.unpack_string()))
+                        }
+                    }
                 }
                 self.alloc(Object::Cmd(new_cmd_parts))
             }

--- a/ipso-eval/src/lib.rs
+++ b/ipso-eval/src/lib.rs
@@ -250,9 +250,9 @@ impl Object {
         }
     }
 
-    pub fn unpack_string(&self) -> &str {
+    pub fn unpack_string(&self) -> Rc<str> {
         match self {
-            Object::String(str) => str,
+            Object::String(str) => str.clone(),
             val => panic!("expected string, got {:?}", val),
         }
     }
@@ -446,7 +446,7 @@ impl Value {
         self.unpack_object().perform_io(interpreter)
     }
 
-    pub fn unpack_string(&self) -> &str {
+    pub fn unpack_string(&self) -> Rc<str> {
         self.unpack_object().unpack_string()
     }
 
@@ -818,7 +818,7 @@ where {
                     |interpreter: &mut Interpreter<'_>, env: Rc<[Value]>, arg: Value| {
                         let a = env[0].unpack_string();
                         let b = arg.unpack_string();
-                        interpreter.alloc_ordering(a.cmp(b))
+                        interpreter.alloc_ordering(a.as_ref().cmp(b.as_ref()))
                     }
                 )
             }
@@ -1081,7 +1081,7 @@ where {
                         let sep = env[0].unpack_string();
                         let s = arg.unpack_string();
                         let a = eval.alloc_values(
-                            s.split(sep)
+                            s.split(sep.as_ref())
                                 .map(|s| eval.alloc(Object::String(Rc::from(s)))),
                         );
                         eval.alloc(Object::Array(a))
@@ -1112,10 +1112,10 @@ where {
                         if strings.is_empty() {
                             eval.alloc(Object::String(Rc::from("")))
                         } else {
-                            let mut joined = String::from(strings[0].unpack_string());
+                            let mut joined = String::from(strings[0].unpack_string().as_ref());
                             for string in &strings[1..] {
-                                joined.push_str(sep);
-                                joined.push_str(string.unpack_string());
+                                joined.push_str(sep.as_ref());
+                                joined.push_str(string.unpack_string().as_ref());
                             }
                             let joined = eval.alloc_str(&joined);
                             eval.alloc(Object::String(joined))
@@ -1160,8 +1160,8 @@ where {
                         let s = arg.unpack_string();
                         let a = eval.alloc_values(
                             Parts {
-                                delimiter: sep,
-                                string: s,
+                                delimiter: sep.as_ref(),
+                                string: s.as_ref(),
                             }
                             .map(|part| eval.alloc(Object::String(Rc::from(part)))),
                         );
@@ -1206,7 +1206,7 @@ where {
                         let a = eval.alloc_values(
                             Partsc {
                                 delimiter: c,
-                                string: s,
+                                string: s.as_ref(),
                             }
                             .map(|part| eval.alloc(Object::String(Rc::from(part)))),
                         );
@@ -1665,7 +1665,7 @@ where {
                     |interpreter: &mut Interpreter, env: Rc<[Value]>, arg: Value| {
                         fn path_exists_io_1(_: &mut Interpreter, env: Rc<[Value]>) -> Value {
                             let path = env[0].unpack_string();
-                            if Path::new(&path).exists() {
+                            if Path::new(path.as_ref()).exists() {
                                 Value::True
                             } else {
                                 Value::False
@@ -1716,7 +1716,7 @@ where {
                     |interpreter: &mut Interpreter, env: Rc<[Value]>, arg: Value| {
                         fn env_getvar_io(interpreter: &mut Interpreter, env: Rc<[Value]>) -> Value {
                             let var = env[0].unpack_string();
-                            match std::env::var(var) {
+                            match std::env::var(var.as_ref()) {
                                 Err(err) => match err {
                                     std::env::VarError::NotPresent => {
                                         // None () : (| None : (), Some : String |)
@@ -1756,7 +1756,7 @@ where {
                             env: Rc<[Value]>,
                         ) -> Value {
                             let var = env[0].unpack_string();
-                            match std::env::var(var) {
+                            match std::env::var(var.as_ref()) {
                                 Err(err) => match err {
                                     std::env::VarError::NotPresent => {
                                         write!(
@@ -1795,7 +1795,7 @@ where {
                         fn env_setvar_io(_: &mut Interpreter, env: Rc<[Value]>) -> Value {
                             let key = env[0].unpack_string();
                             let value = env[1].unpack_string();
-                            std::env::set_var(key, value);
+                            std::env::set_var(key.as_ref(), value.as_ref());
                             Value::Unit
                         }
                         let env = interpreter.alloc_values({
@@ -1859,8 +1859,8 @@ where {
                 |interpreter: &mut Interpreter<'_>, _: Rc<[Value]>, arg: Value| {
                     fn file_read_io(interpreter: &mut Interpreter<'_>, env: Rc<[Value]>) -> Value {
                         let path = env[0].unpack_string();
-                        let contents =
-                            std::fs::read_to_string(path).unwrap_or_else(|err| panic!("{}", err));
+                        let contents = std::fs::read_to_string(path.as_ref())
+                            .unwrap_or_else(|err| panic!("{}", err));
                         interpreter.alloc(Object::String(Rc::from(contents)))
                     }
 
@@ -1879,7 +1879,8 @@ where {
                         fn file_write_io(_: &mut Interpreter, env: Rc<[Value]>) -> Value {
                             let path = env[0].unpack_string();
                             let content = env[1].unpack_string();
-                            std::fs::write(path, content).unwrap_or_else(|err| panic!("{}", err));
+                            std::fs::write(path.as_ref(), content.as_ref())
+                                .unwrap_or_else(|err| panic!("{}", err));
                             Value::Unit
                         }
                         let env = interpreter.alloc_values({
@@ -1906,7 +1907,7 @@ where {
                             let mut file = std::fs::OpenOptions::new()
                                 .create(true)
                                 .append(true)
-                                .open(path)
+                                .open(path.as_ref())
                                 .unwrap_or_else(|err| panic!("{}", err));
                             file.write_all(content.as_bytes())
                                 .unwrap_or_else(|err| panic!("{}", err));
@@ -2134,7 +2135,7 @@ where {
                 StringPart::Expr(expr) => {
                     let s = self.eval(env, expr);
                     let s = s.unpack_string();
-                    value.push_str(s);
+                    value.push_str(s.as_ref());
                 }
                 StringPart::String(s) => value.push_str(s.as_str()),
             }
@@ -2507,7 +2508,7 @@ where {
                                 }
                                 Pattern::String(actual_string) => {
                                     let expected_string = expr.unpack_string();
-                                    if expected_string == actual_string.as_ref() {
+                                    if expected_string.as_ref() == actual_string.as_ref() {
                                         target = Some(&branch.body);
                                         break;
                                     }

--- a/ipso-import/src/lib.rs
+++ b/ipso-import/src/lib.rs
@@ -302,10 +302,13 @@ pub fn rewrite_module_accessors_expr(
                 exclude.remove_all(arg_names.into_iter());
             }
         }
-        syntax::Expr::Cmd(parts) => parts.iter_mut().for_each(|part| {
-            part.value.iter_mut().for_each(|string_part| {
+        syntax::Expr::Cmd(cmd_parts) => cmd_parts.iter_mut().for_each(|cmd_part| match cmd_part {
+            syntax::CmdPart::Arg(string_parts) => string_parts.iter_mut().for_each(|string_part| {
                 rewrite_module_accessors_string_part(exclude, imported_items, string_part)
-            })
+            }),
+            syntax::CmdPart::Args(expr) => {
+                rewrite_module_accessors_expr(exclude, imported_items, expr)
+            }
         }),
         syntax::Expr::Comp(_) => {
             /*

--- a/ipso-lex/src/lib.rs
+++ b/ipso-lex/src/lib.rs
@@ -713,9 +713,7 @@ impl<'input> Iterator for Lexer<'input> {
                             self.consume();
 
                             match self.current {
-                                Some(c) if c == '{' => {
-                                    let pos = self.pos;
-                                    let column = self.column;
+                                Some('{') => {
                                     self.consume();
 
                                     self.context.push(Mode::Normal);
@@ -725,6 +723,48 @@ impl<'input> Iterator for Lexer<'input> {
                                         pos,
                                         column,
                                     })
+                                }
+                                Some('.') => {
+                                    self.consume();
+
+                                    match self.current {
+                                        Some('.') => {
+                                            self.consume();
+
+                                            match self.current {
+                                                Some('{') => {
+                                                    self.consume();
+
+                                                    self.context.push(Mode::Normal);
+
+                                                    Some(Token {
+                                                        data: token::Data::DollarDotDotLBrace,
+                                                        pos,
+                                                        column,
+                                                    })
+                                                }
+                                                _ => {
+                                                    self.context.push(Mode::Ident);
+
+                                                    Some(Token {
+                                                        data: token::Data::DollarDotDot,
+                                                        pos,
+                                                        column,
+                                                    })
+                                                }
+                                            }
+                                        }
+                                        Some(c) => Some(Token {
+                                            data: token::Data::Unexpected(c),
+                                            pos: self.pos,
+                                            column: self.column,
+                                        }),
+                                        None => Some(Token {
+                                            data: token::Data::Unexpected('.'),
+                                            pos: self.pos,
+                                            column: self.column,
+                                        }),
+                                    }
                                 }
                                 _ => {
                                     self.context.push(Mode::Ident);

--- a/ipso-lex/src/test.rs
+++ b/ipso-lex/src/test.rs
@@ -918,3 +918,98 @@ fn lex_cmd_4() {
     let actual = lexer.collect::<Vec<Token>>();
     assert_eq!(expected, actual)
 }
+
+#[test]
+fn lex_cmd_5() {
+    let input = Rc::from("`echo $..a`");
+    let lexer = Lexer::new(&input);
+    let expected = vec![
+        Token {
+            data: token::Data::Backtick,
+            pos: 0,
+            column: 0,
+        },
+        Token {
+            data: token::Data::Cmd(Rc::from("echo")),
+            pos: 1,
+            column: 1,
+        },
+        Token {
+            data: token::Data::Space,
+            pos: 5,
+            column: 5,
+        },
+        Token {
+            data: token::Data::DollarDotDot,
+            pos: 6,
+            column: 6,
+        },
+        Token {
+            data: token::Data::Ident(Rc::from("a")),
+            pos: 9,
+            column: 9,
+        },
+        Token {
+            data: token::Data::Backtick,
+            pos: 10,
+            column: 10,
+        },
+        Token {
+            data: token::Data::Eof,
+            pos: 11,
+            column: 11,
+        },
+    ];
+    let actual = lexer.collect::<Vec<Token>>();
+    assert_eq!(expected, actual)
+}
+
+#[test]
+fn lex_cmd_6() {
+    let input = Rc::from("`echo $..{a}`");
+    let lexer = Lexer::new(&input);
+    let expected = vec![
+        Token {
+            data: token::Data::Backtick,
+            pos: 0,
+            column: 0,
+        },
+        Token {
+            data: token::Data::Cmd(Rc::from("echo")),
+            pos: 1,
+            column: 1,
+        },
+        Token {
+            data: token::Data::Space,
+            pos: 5,
+            column: 5,
+        },
+        Token {
+            data: token::Data::DollarDotDotLBrace,
+            pos: 6,
+            column: 6,
+        },
+        Token {
+            data: token::Data::Ident(Rc::from("a")),
+            pos: 10,
+            column: 10,
+        },
+        Token {
+            data: token::Data::RBrace,
+            pos: 11,
+            column: 11,
+        },
+        Token {
+            data: token::Data::Backtick,
+            pos: 12,
+            column: 12,
+        },
+        Token {
+            data: token::Data::Eof,
+            pos: 13,
+            column: 13,
+        },
+    ];
+    let actual = lexer.collect::<Vec<Token>>();
+    assert_eq!(expected, actual)
+}

--- a/ipso-lex/src/token/mod.rs
+++ b/ipso-lex/src/token/mod.rs
@@ -31,6 +31,8 @@ pub enum Name {
     DoubleQuote,
     Dollar,
     DollarLBrace,
+    DollarDotDot,
+    DollarDotDotLBrace,
     String,
     SingleQuote,
     Char,
@@ -80,6 +82,8 @@ impl Arbitrary for Name {
             Name::DoubleQuote,
             Name::Dollar,
             Name::DollarLBrace,
+            Name::DollarDotDot,
+            Name::DollarDotDotLBrace,
             Name::String,
             Name::SingleQuote,
             Name::Char,
@@ -121,7 +125,7 @@ impl Arbitrary for Name {
 
 impl Name {
     pub fn num_variants() -> usize {
-        44 + Keyword::num_variants()
+        46 + Keyword::num_variants()
     }
 
     pub fn from_int(ix: usize) -> Option<Self> {
@@ -187,6 +191,8 @@ impl Name {
             59 => Some(Self::LParenPipe),
             60 => Some(Self::PipeRParen),
             61 => Some(Self::Space),
+            62 => Some(Self::DollarDotDot),
+            63 => Some(Self::DollarDotDotLBrace),
             _ => None,
         }
     }
@@ -254,6 +260,8 @@ impl Name {
             Self::LParenPipe => 59,
             Self::PipeRParen => 60,
             Self::Space => 61,
+            Self::DollarDotDot => 62,
+            Self::DollarDotDotLBrace => 63,
         }
     }
 
@@ -266,6 +274,8 @@ impl Name {
             Name::DoubleQuote => String::from("'\"'"),
             Name::Dollar => String::from("'$'"),
             Name::DollarLBrace => String::from("'${'"),
+            Name::DollarDotDot => String::from("'$..'"),
+            Name::DollarDotDotLBrace => String::from("'$..{'"),
             Name::String => String::from("string"),
             Name::SingleQuote => String::from("'"),
             Name::Char => String::from("character"),
@@ -339,6 +349,8 @@ pub enum Data {
     DoubleQuote,
     Dollar,
     DollarLBrace,
+    DollarDotDot,
+    DollarDotDotLBrace,
     String {
         value: String,
         length: usize,
@@ -432,6 +444,8 @@ impl Data {
             Data::SingleQuote => 1,
             Data::Dollar => 1,
             Data::DollarLBrace => 2,
+            Data::DollarDotDot => 3,
+            Data::DollarDotDotLBrace => 4,
             Data::String { length, .. } => *length,
             Data::Char { length, .. } => *length,
             Data::Pipe => 1,
@@ -459,6 +473,8 @@ impl Data {
             Data::DoubleQuote => Name::DoubleQuote,
             Data::Dollar => Name::Dollar,
             Data::DollarLBrace => Name::DollarLBrace,
+            Data::DollarDotDot => Name::DollarDotDot,
+            Data::DollarDotDotLBrace => Name::DollarDotDotLBrace,
             Data::String { .. } => Name::String,
             Data::SingleQuote => Name::SingleQuote,
             Data::Char { .. } => Name::Char,

--- a/ipso-parse/src/grammar/expr/mod.rs
+++ b/ipso-parse/src/grammar/expr/mod.rs
@@ -5,7 +5,7 @@ mod test;
 
 use crate::{
     between, choices, grammar::pattern::pattern, indent, indent_scope, keep_left, keep_right, many,
-    map0, map2, operator::operator, optional, sep_by, spanned, Parsed, Parser,
+    many1, map0, map2, operator::operator, optional, sep_by, spanned, Parsed, Parser,
 };
 use ipso_lex::token;
 use ipso_syntax::{Binop, Branch, CmdPart, CompLine, Expr, Keyword, Spanned, StringPart};
@@ -257,6 +257,8 @@ pub fn expr_array(parser: &mut Parser) -> Parsed<Expr> {
 cmd_part ::=
   '"' string_part* '"'
   ( cmd_char* | '$' ident | '${' expr '}' )+
+  '$..' ident
+  '$..{' expr '}'
 ```
 
 ```text
@@ -277,15 +279,25 @@ pub fn cmd_part(parser: &mut Parser) -> Parsed<CmdPart> {
                 string_part_expr(parser),
                 string_part_string(parser)
             ))
-        ),
-        many!(choices!(
+        )
+        .map(CmdPart::Arg),
+        many1!(choices!(
             string_part_expr(parser),
             parser
                 .cmd_part_literal()
                 .map(|string| StringPart::String(String::from(string.as_ref())))
         ))
+        .map(CmdPart::Arg),
+        between!(
+            parser.token(&token::Data::DollarDotDotLBrace),
+            parser.token(&token::Data::RBrace),
+            expr(parser).map(CmdPart::Args)
+        ),
+        keep_right!(
+            parser.token(&token::Data::DollarDotDot),
+            spanned!(parser, parser.ident_owned().map(Expr::Var)).map(CmdPart::Args)
+        )
     )
-    .map(|value| CmdPart { value })
 }
 
 /**

--- a/ipso-parse/src/lib.rs
+++ b/ipso-parse/src/lib.rs
@@ -417,6 +417,14 @@ macro_rules! many {
 }
 
 #[macro_export]
+macro_rules! many1 {
+    ($x:expr) => {{
+        use $crate::many_with;
+        $x.and_then(|x| many_with!(vec![x], $x))
+    }};
+}
+
+#[macro_export]
 macro_rules! choices {
     ($x:expr, $y:expr) => {{
         use $crate::Parsed;

--- a/ipso-repl/src/lib.rs
+++ b/ipso-repl/src/lib.rs
@@ -153,7 +153,7 @@ impl Repl {
         stdout: &mut dyn Write,
         stderr: &mut dyn Write,
         expr: Spanned<ipso_syntax::Expr>,
-    ) -> Result<Option<String>, Error> {
+    ) -> Result<Option<Rc<str>>, Error> {
         let mut expr = desugar_expr(&self.source, expr)?;
         rewrite_module_accessors_expr(&mut Default::default(), &self.imported_items, &mut expr);
 
@@ -301,11 +301,11 @@ impl Repl {
                 | Object::Closure { .. }
                 | Object::StaticClosure { .. }
                 | Object::Cmd(_) => todo!(),
-                Object::String(s) => Ok(Some(String::from(s.as_ref()))),
+                Object::String(s) => Ok(Some(s.clone())),
                 Object::IO { env, body } => {
                     let result = body.run(&mut interpreter, env.clone());
                     Ok(if show_final_value {
-                        Some(String::from(result.unpack_string()))
+                        Some(result.unpack_string())
                     } else {
                         // `show_final_value` should only be false for `IO ()`
                         debug_assert!(result == ipso_eval::Value::Unit);

--- a/ipso-syntax/src/desugar/mod.rs
+++ b/ipso-syntax/src/desugar/mod.rs
@@ -134,10 +134,12 @@ fn desugar_cmd_part_mut(
     var_gen: &mut VarGen,
     cmd_part: &mut CmdPart,
 ) -> Result<(), Error> {
-    cmd_part
-        .value
-        .iter_mut()
-        .try_for_each(|string_part| desugar_string_part_mut(source, var_gen, string_part))
+    match cmd_part {
+        CmdPart::Arg(string_parts) => string_parts
+            .iter_mut()
+            .try_for_each(|string_part| desugar_string_part_mut(source, var_gen, string_part)),
+        CmdPart::Args(expr) => desugar_expr_mut(source, var_gen, expr),
+    }
 }
 
 fn desugar_branches_mut(

--- a/ipso-syntax/src/lib.rs
+++ b/ipso-syntax/src/lib.rs
@@ -378,8 +378,9 @@ echo $a/$b
 ```
 */
 #[derive(Debug, PartialEq, Eq, Clone)]
-pub struct CmdPart {
-    pub value: Vec<StringPart>,
+pub enum CmdPart {
+    Arg(Vec<StringPart>),
+    Args(Spanned<Expr>),
 }
 
 #[derive(Debug, PartialEq, Eq, Clone, Copy)]

--- a/ipso-typecheck/src/type_inference/mod.rs
+++ b/ipso-typecheck/src/type_inference/mod.rs
@@ -953,13 +953,23 @@ pub fn check(
 
             let cmd_parts = cmd_parts
                 .iter()
-                .map(|cmd_part| {
-                    let value = cmd_part
-                        .value
-                        .iter()
-                        .map(|string_part| check_string_part(env, state, string_part))
-                        .collect::<Result<Vec<_>, _>>()?;
-                    Ok(CmdPart { value })
+                .map(|cmd_part| match cmd_part {
+                    syntax::CmdPart::Arg(string_parts) => {
+                        let string_parts = string_parts
+                            .iter()
+                            .map(|string_part| check_string_part(env, state, string_part))
+                            .collect::<Result<Vec<_>, _>>()?;
+                        Ok(CmdPart::Arg(string_parts))
+                    }
+                    syntax::CmdPart::Args(expr) => {
+                        let expr = check(
+                            env,
+                            state,
+                            expr,
+                            &Type::app(Type::mk_array(env.common_kinds), Type::String),
+                        )?;
+                        Ok(CmdPart::Args(expr))
+                    }
                 })
                 .collect::<Result<_, Error>>()?;
 

